iface/get.go

```diff
@@ -71,6 +71,19 @@ func GetInterface(dir, ifaceName string) (Interface, error) {
 		Package: pkg.Name,
 		Name:    ifaceObj.Name(),
 	}
+	qualifier := Qualify(pkg.Types, imps, &iface.Imports)
+
+	// Record type parameter list info.
+	if ifaceNamed, ok := ifaceObj.Type().(*types.Named); ok {
+		typeParams := ifaceNamed.TypeParams()
```

Comment:
There is a `validateType` function in this file to recursively validate all the types involved in the interface. This function includes the case where the [type is `*types.Named`](https://github.com/nicheinc/mock/blob/a68fb3801a78dfbd7e0c4d8cc294779929fd84ff/iface/get.go#L297-L302), and I considered validating the type constraints at that point. However, the error log is currently commented out, and the default case, which in theory is the only way this function can return "not valid", instead returns "valid".

So in short, this function doesn't do anything - it will always just return `true`, for any type. I don't understand the code well enough to be comfortable removing it entirely, and changing the `default` case to return "not valid" causes the well-formed example types to fail. 😞 So I'm just going to leave it alone for now.

Reply:
> the default case, which in theory is the only way this function can return "not valid", instead returns "valid".

It's been awhile since I've looked at the code, but I think [this](https://github.com/nicheinc/mock/blob/a68fb3801a78dfbd7e0c4d8cc294779929fd84ff/iface/get.go#L251) is actually the main way the function can return "not valid".

Reply:
Wow, I don't know how I glossed over that case. :see_no_evil: Maybe I'll try to incorporate those constraint types after all then. Also, hey Cochran! :wave:

Reply:
Hey!! 👋

Looking forward to the PRs for merging these changes upstream 😉😉😉

Reply:
And just for clarity, the purpose of `ValidateType` is to check whether any of the type errors in the package are actually related to the interface you're trying to mock. It makes it possible to generate mocks for an interface even if there are other unrelated errors in the package, and only prevents mock generation if any of the errors are actually relevant to the interface definition.

Originally, the tool would error out and fail to generate a mock if `len(pkg.Errors) > 0`, but that meant you couldn't generate a mock if you had _any_ errors in the package. I often wanted to (re-)generate a mock when I was in the middle of work (and other stuff was broken), so I changed it to stop erroring out when `len(pkg.Errors) > 0` and instead added `ValidateType`, so that  I could determine if any of the types actually involved in the interface were broken/invalid, and only fail to generate a mock in that case (but it still [returns all of the package errors](https://github.com/nicheinc/mock/blob/a68fb3801a78dfbd7e0c4d8cc294779929fd84ff/iface/get.go#L63), when that happens).

`ValidateType` just recursively descends from more "complex" types to the "basic" types that constitute them (for example, from `*chan[int]` to `chan[int]` to `int`). When it gets to the bottom level, it checks whether the type is `types.Invalid`, which happens if the type isn't defined. Like, I believe you'd hit that case for `chan[Invalid]` if `Invalid` wasn't defined.

Reply:
Awesome, thanks for that explanation! I think we'd definitely want to include the type parameter constraints in this check then since they need to be repeated in the mocked struct definition, and we wouldn't want to generate a type with broken constraints. :+1:

Reply:
Agreed!